File: kinosan.rb
# time_start_ex = Time.now
SL = 31		#分割する桁数（とりあえず32ビットの正の数分=31桁）

h, w = gets.split(' ').map(&:to_i)
lines = $stdin.read().split(?\n).map(&:to_s)
home_lines = lines[0,h]
widgets = lines[h+1,lines[h].to_i]

# time_end = Time.now
# ptime = (time_end - time_start_ex) * 1000;
# puts '!!!STDIN_INPUT_TIME->!!!' << ptime.to_s() << "\n";

if w > SL
	#ホームの幅が広い場合はSL幅で計算
	ww = SL
else
	ww = w
end

bin_max_val = 2 ** ww -1
str_max_val = '1' * ww

#衝突チェックブロック用ハッシュ
check_block = Hash.new {|hash, key| hash[key] = Hash.new()}
#スキマチェック用ハッシュ
check_list = Hash.new {|hash, key| hash[key] = Hash.new()}

sort_cond = h - 1
count_max_sh = 0
next_max_sh = h
home_lines.each_with_index {|hl_val, j|
	#[最初の準備] ホームの情報をSLの幅で分けて2進数にし、配列に格納
	bin_home_list = Array.new
	pos = 0
	while pos < w do
		work_home = hl_val[pos,SL]
		if work_home.length < ww
			#最後のブロックの右側を1で埋める
			work_home = work_home.ljust(ww,'1')
		end
		bin_home_list.push(work_home.to_i(2))		#2進数に変換して格納
		pos += SL
	end

	#puts bin_home_list

	sh = j + 1
	while sh > 0 do
		if sh<=next_max_sh
			# [手順1] ウィジェットの高さパターンに応じてOR演算をしたリストを作成
			#(例)
			#     00100000
			#     01000011　⇒ 01100111
			#     01100100
			str_check_list = ''
			if sh>1
				sh_ck_block = check_block[sh-1]
			end
			sh_ck_block_new = check_block[sh]
			bin_home_list.each_with_index {|val, i|
				if sh==1
					chk = val
				else
					#OR演算
					if sh_ck_block[i]==bin_max_val
						chk = bin_max_val
					else
						chk = sh_ck_block[i] | val
					end
				end
				sh_ck_block_new[i] = chk
				#puts check_block

				#求めた2進数を文字列に変換
				if chk == bin_max_val
					str_check_list << str_max_val
				else
					wk_str = chk.to_s(2)
					if wk_str.length < ww
						wk_str = wk_str.rjust(ww,'0')
					end 
					str_check_list << wk_str
				end
			}

			#puts str_check_list

			# [手順2] 求めた2進数からゼロ文字列の長さリストを作成
			#(例)
			#     0011000110101011　⇒　2,3,1,1,1
			pos = 0
			sh_ck_list = check_list[sh]
			while pos < w do
				first = str_check_list.index('0',pos)
				if first.nil?
					if pos == 0
						#ゼロが存在しない　→　次のループより入りきるウィジェット高さが制限される
						next_max_sh = sh
					end
					break
				else
					second = str_check_list.index('1',first)
					if second.nil?
						key = w - first
						pos = w
					else
						key = second - first
						pos = second
					end
					if sh_ck_list.key?(key)
						sh_ck_list[key]+=1
					else
						sh_ck_list[key]=1
					end
				end
			end
			if j == sort_cond
				if sh_ck_list.length > 1
					check_list[sh] = sh_ck_list.sort_by{|key,val| -key}
					if sh>count_max_sh
						count_max_sh=sh
					end
				end
			end
			sh -= 1
		else
			sh = next_max_sh
		end
	end
}

#puts check_list

# time_end = Time.now
# ptime = (time_end - time_start_ex) * 1000;
# puts '!!!PREPARED_TIME->!!!' << ptime.to_s() << "\n";

puts_str = '';
for w_val in widgets do
	#ウィジェットループ
	sh, tw = w_val.split(' ').map(&:to_i)
	cnt = 0
	if sh <= count_max_sh
		check_list[sh].each{|key,value|
			break if key < tw
			cnt += (key - tw + 1) * value
		}
	end
	puts_str << cnt.to_s << "\n"
end
puts puts_str
# time_end = Time.now
# ptime = (time_end - time_start_ex) * 1000;
# puts '!!!TOTAL_TIME->!!!' << ptime.to_s() << "\n";

File: kinosan.py
# -*- coding: utf-8 -*-
# python 2.7.3
# import sys,time
import sys
# time_start_ex = time.time()

SL=31		#分割する桁数（とりあえず32ビットの正の数分=31桁）

# def decbin(bin,w):
#     if bin == 0:
# 		return '0' * w
#     bin_str = ''
#     i = w - 1
#     while i >= 0:
# 		bi = (bin >> i) & 1
# 		bin_str += str(bi)
# 		i -= 1
#     return bin_str

(H,W) = map(int, sys.stdin.readline().split(' '))
ALL = map(str, sys.stdin.read().splitlines())
home_lines = ALL[0:H]
N = int(ALL[H])
widgets = ALL[H+1:]
#print '!!!check!!!'
#print home_lines
#print widgets

# time_end = time.time()
# ptime = time_end - time_start_ex
# ptime = ptime * 1000
# ptime = str(ptime)
# print "!!!INPUT_TIME->" + ptime + "!!!"


if W > SL:
	WW = SL
else:
	WW = W

bin_max_val = pow(2,WW) - 1		#2進数最大値 
str_max_val = '1' * WW

#衝突チェックブロック配列
block_cnt = W // SL + 1
check_block = [[0 for j in range(block_cnt)] for i in range(H + 1)];

#スキマチェック用リスト
check_list = [{} for i in range(H + 1)];

#print check_block
#print check_list

SORT_COND = H - 1
for j,hl_val in enumerate(home_lines):

	#[最初の準備] ホームの情報をSLの幅で分けて2進数にし、配列に格納
	bin_home_list = []
	pos = 0
	append = bin_home_list.append
	while pos<W:
		work_home = hl_val[pos:pos+SL]
		if len(work_home)<WW:
			work_home = work_home.ljust(WW,'1')
		append(int(work_home,2))
		pos += SL

	SH = j + 1
	while SH > 0:
		# [手順1] ウィジェットの高さパターンに応じてOR演算をしたリストを作成
		#(例)
		#     00100000
		#     01000011　⇒ 01100111
		#     01100100
		#str_check_list = '';
		str_list = []
		append = str_list.append
		cblock = check_block[SH - 1]
		cblock_now = check_block[SH]
		for i,val in enumerate(bin_home_list):
			if SH==1:
				chk = val
			else:
				if cblock[i] == bin_max_val:
					chk = bin_max_val
				else:
					chk = cblock[i] | val
			cblock_now[i] = chk;

			#求めた2進数を文字列に変換
			if chk == bin_max_val:
				append(str_max_val)
			else:
				wk_str = bin(chk)
				wk_str = wk_str[2:].rjust(WW,'0')
				append(wk_str)

		# [手順2] 求めた2進数からゼロ文字列の長さリストを作成
		#(例)
		#     00100010101011　⇒　2,3,1,1,1
		str_check_list = "".join(str_list)
		clist = check_list[SH]
		pos = 0
		while pos < W:
			first = str_check_list.find('0',pos)
			if first >= 0:
				second = str_check_list.find('1',first)
				if second >= 0:
					key = second - first
					pos = second
				else:
					key = W - first
					pos = W
				if clist.has_key(key):
					clist[key] += 1
				else:
					clist[key] = 1
			else:
				break
		if j==SORT_COND:
			if len(clist) > 0:
			 	check_list[SH] = sorted(clist.items(),key=lambda x: x[0], reverse=True)
		SH -= 1

# time_end = time.time()
# ptime = time_end - time_start_ex
# ptime = ptime * 1000
# ptime = str(ptime)
# print "!!!PREPARED_TIME->" + ptime + "!!!"

#print check_list
#print_str = '';
print_list = ['0' for n in range(N)]
index=0
for index,w_val in enumerate(widgets):
	(SH,TW) = map(int, w_val.split(' '))
	cnt = 0
	if SH <= H:
		for key,val in check_list[SH]:
			if key < TW:
				break
			cnt += (key-TW+1) * val
		if cnt > 0:
			print_list[index] = str(cnt)
print "\n".join(print_list)
# time_end = time.time()
# ptime = time_end - time_start_ex
# ptime = ptime * 1000
# ptime = str(ptime)
# print "!!!TOTAL_TIME->" + ptime + "!!!"